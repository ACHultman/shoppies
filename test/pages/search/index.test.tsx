import { getPage } from 'next-page-tester'
import React from 'react'
import Search from '../../../pages/search'
import { render, waitFor } from '../../testUtils'
import { screen, fireEvent } from '@testing-library/dom'
import Router from 'next/router'

jest.mock('next/router', () => {
  return {
    __esModule: true,
    ...jest.requireActual<Record<string, unknown>>('next/router'),
    default: {},
    push: jest.fn(),
  }
})

describe('Search page', () => {
  beforeEach(async () => {
    const { render } = await getPage({
      route: '/search',
      router: (router) => Object.assign(Router, router),
    })
    render()
  })

  it('matches snapshot', () => {
    const { asFragment } = render(<Search />, {})
    expect(asFragment()).toMatchSnapshot()
  })

  it('handles search input', () => {
    const textbox: HTMLInputElement = screen.getByLabelText(
      'search-input'
    ) as HTMLInputElement
    fireEvent.change(textbox, {
      target: { value: 'test' },
    })

    expect(textbox.value).toBe('test')
  })

  it('alerts upon empty search', async () => {
    const textbox: HTMLInputElement = screen.getByLabelText(
      'search-input'
    ) as HTMLInputElement
    fireEvent.change(textbox, {
      target: { value: '' },
    })
    expect(textbox.value).toBe('')

    const submitButton = screen.getByLabelText('search-submit')
    fireEvent.click(submitButton)
    expect(
      screen.getByText(
        /Enter a name of your favourite movie \(e\.g\. "tenet"\)\./i
      )
    ).toBeInTheDocument()
  })

  it('alerts when search terms are less than three characters long', async () => {
    const textbox: HTMLInputElement = screen.getByLabelText(
      'search-input'
    ) as HTMLInputElement
    fireEvent.change(textbox, {
      target: { value: '12' },
    })
    expect(textbox.value).toBe('12')

    const submitButton = screen.getByLabelText('search-submit')
    fireEvent.click(submitButton)
    expect(
      screen.getByText(
        /Hint: Your search should be at least three characters long\./i
      )
    ).toBeInTheDocument()
  })

  it('routes to search page for entered search terms', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({ Search: [] }),
      })
    ) as any

    const textbox: HTMLInputElement = screen.getByLabelText(
      'search-input'
    ) as HTMLInputElement
    fireEvent.change(textbox, {
      target: { value: 'test' },
    })
    expect(textbox.value).toBe('test')

    const submitButton = screen.getByLabelText('search-submit')
    fireEvent.click(submitButton)
    await waitFor(() => screen.getByText(/Sorry! No results found for/i))
  })
})
