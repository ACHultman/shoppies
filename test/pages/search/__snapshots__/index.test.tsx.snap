// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto pt-6"
  >
    <div
      class="flex justify-center max-w-screen-sm mx-auto overflow-hidden px-10"
    >
      <form
        class="w-full h-10 pl-3 pr-2 bg-white border rounded-full flex justify-between items-center relative"
      >
        <input
          autocomplete="off"
          class="text-gray-800 appearance-none w-full outline-none focus:outline-none active:outline-none"
          name="query"
          placeholder="Search movies by name..."
          type="text"
          value=""
        />
        <button
          class="ml-1 outline-none focus:outline-none active:outline-none"
          type="submit"
        >
          <svg
            class="w-6 h-6 fill-current text-gray-300"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
            />
          </svg>
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
